### PR TITLE
Consistent quote wrapping around MessageBuilder.format

### LINKED ISSUE #3409: MessageBuilder.format inconsistently surrounds types with quotes
When working on #3402, using `msg.format` for types usually surrounded them with quotes, e.g., in test `testUnionAttributeAccess`, calling `msg.format` while printing generates the following:
```
Element "C" of "Union[A, C]" has no attribute "y"
```
However, in one test case, ``, the type is not surrounded by quotes (see last line, I attached the entire test case for context):

```
 [case testGenericTypeAliasesUnion]
 from typing import TypeVar, Generic, Union, Any
 T = TypeVar('T')
 class Node(Generic[T]):
     def __init__(self, x: T) -> None:
         self.x = x
 
 UNode = Union[int, Node[T]]
 x = 1 # type: UNode[int]
 
 x + 1 # E: Unsupported left operand type for + (some union)
 if not isinstance(x, Node):
     x + 1
 
 if not isinstance(x, int):
    x.x = 1
    x.x = 'a' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
 
 def f(x: T) -> UNode[T]:
     if 1:
         return Node(x)
     else:
         return 1
 
 reveal_type(f(1)) # E: Revealed type is 'Union[builtins.int, __main__.Node[builtins.int*]]'
 
 TNode = Union[T, Node[int]]
 s = 1 # type: TNode[str] # E: Incompatible types in assignment (expression has type "int", variable has type "Union[str, Node[int]]")
 
 if not isinstance(s, str):
     s.x = 1
  
  z = None # type: TNode # Same as TNode[Any]
  z.x
  z.foo() # E: Element Node[int] of "Union[Any, Node[int]]" has no attribute "foo"
```
Notice how `Node[int]` is not surrounded by quotes. The relevant `format` function is defined in `mypy/messages.py`.

(Sorry for only reproducing this in an unrelated test, I don't have enough knowledge on the code base to  create the relevant type object directly).

### PR DESCRIPTION
**Pull request by OddBloke**
This is a refreshed version of #3430, and fixes #3409.

### CODE REVIEW

File: test-data/unit/check-varargs.test [hunk 1]
```diff
@@ -277,13 +277,13 @@ class A: pass
 class B: pass
 [builtins fixtures/list.pyi]
 [out]
-main:3: error: Argument 1 to "f" has incompatible type *List[A]; expected "B"
-main:4: error: Argument 2 to "f" has incompatible type *List[A]; expected "B"
+main:3: error: Argument 1 to "f" has incompatible type *"List[A]"; expected "B"
```
Comment (ilevkivskyi) [inline test-data/unit/check-varargs.test]: I think the star should be inside quotes for types like this.

**Comment by OddBloke** [comment]
Yep, agreed. Away from my laptop for the rest of the day, so I'll look at
it tomorrow.

**Comment by OddBloke** [comment]
So looking at how the formatting of the star-args is done, I think the formatting methods should probably be refactored a bit.

Currently, the "public" API (`.format` and `.format_simple`)  will only ever return quoted strings, which mean that callers have to remove the quotes on those strings if they want to use them for something else (like, for example, prepending * or ** inside the quotes). I think we should have additional methods that will return the "bare" formatted string (which the existing methods can then just wrap with the quoting logic).

Also, as far as I can tell, the only way that `.format_simple` gets called in the mypy code base is by `.format` (or one of the things that only `.format` calls), so I'll also look at dropping `.format_simple` in favour of just `._format_simple`.

**Comment by OddBloke** [comment]
I've done that refactoring and addressed the comment from @ilevkivskyi.

File: test-data/unit/check-protocols.test [hunk 2]
```diff
@@ -1663,8 +1663,8 @@ def apply_gen(f: Callable[[T], T]) -> T:
 reveal_type(apply_gen(Add5())) # E: Revealed type is 'builtins.int*'
 def apply_str(f: Callable[[str], int], x: str) -> int:
     return f(x)
-apply_str(Add5(), 'a') # E: Argument 1 to "apply_str" has incompatible type "Add5"; expected Callable[[str], int] \
-                       # N: 'Add5.__call__' has type 'Callable[[Arg(int, 'x')], int]'
+apply_str(Add5(), 'a') # E: Argument 1 to "apply_str" has incompatible type "Add5"; expected "Callable[[str], int]" \
+                       # N: 'Add5.__call__' has type "Callable[[Arg(int, 'x')], int]"
```
Comment (ilevkivskyi) [inline test-data/unit/check-protocols.test]: Another potential idea: maybe use double quotes instead of single quotes in ``'Add5.__call__'``? (But not in this PR.)

File: test-data/unit/check-tuples.test [hunk 3]
```diff
@@ -579,7 +579,7 @@ class LongTypeName:
     def __add__(self, x: 'LongTypeName') -> 'LongTypeName': pass
 [builtins fixtures/tuple.pyi]
 [out]
-main:3: error: Unsupported operand types for + ("LongTypeName" and tuple(length 50))
+main:3: error: Unsupported operand types for + ("LongTypeName" and "tuple(length 50)")
```
Comment (ilevkivskyi) [inline test-data/unit/check-tuples.test]: I think ``tuple(length 50)`` should not be surrounded by quotes.

File: mypy/messages.py [hunk 4]
```diff
@@ -184,69 +184,27 @@ def warn(self, msg: str, context: Context, file: Optional[str] = None,
...
-                return 'Callable[[{}], {}]'.format(", ".join(arg_strings), return_type)
-            else:
-                # Use a simple representation for function types; proper
-                # function types may result in long and difficult-to-read
-                # error messages.
-                return 'overloaded function'
-        else:
-            # Default case; we simply have to return something meaningful here.
-            return 'object'
+        Convert a type to a relatively short string suitable for error messages.
```
[the remark this answers is not in the document]
Comment (OddBloke) [reply mypy/messages.py]: Line 189 is blank, it's just an existing blank line that the diff has kept from somewhere.

File: mypy/messages.py [hunk 5]
```diff
@@ -184,69 +184,27 @@ def warn(self, msg: str, context: Context, file: Optional[str] = None,
...
+        if ret in ['Module', 'overloaded function']:
+            # Messages are easier to read if these aren't quoted
+            return ret
+        return '"{}"'.format(ret)
 
-        Return "" for complex types. Try to keep the length of the result
-        relatively short to avoid overly long error messages.
+    def format_bare(self, typ: Type, verbosity: int = 0) -> str:
+        """
+        Convert a type to a relatively short string suitable for error messages.
```
[the remark this answers is not in the document]
Comment (OddBloke) [reply mypy/messages.py]: Same as before, there is a blank line it's just not considered added for whatever reason.

File: mypy/messages.py [hunk 6]
```diff
@@ -340,25 +294,67 @@ def format_simple(self, typ: Type, verbosity: int = 0) -> str:
...
+            return 'object'
 
-    def format_distinctly(self, type1: Type, type2: Type) -> Tuple[str, str]:
+    def format_distinctly(self, type1: Type, type2: Type, bare: bool = False) -> Tuple[str, str]:
         """Jointly format a pair of types to distinct strings.
 
         Increase the verbosity of the type strings until they become distinct.
         """
+        if bare:
+            format_method = self.format_bare
```
[the remark this answers is not in the document]
Comment (OddBloke) [reply mypy/messages.py]: I've expanded the docstring.

File: test-data/unit/check-async-await.test [hunk 7]
```diff
@@ -312,7 +312,7 @@ class C:
     async def __aenter__(self) -> int: pass
     def __aexit__(self, x, y, z) -> int: pass
 async def f() -> None:
-    async with C() as x: # E: Incompatible types in "async with" for __aexit__ (actual type "int", expected type Awaitable[Any])
+    async with C() as x: # E: Incompatible types in "async with" for __aexit__ (actual type "int", expected type "Awaitable[Any]")
```
[the remark this answers is not in the document]
Comment (OddBloke) [reply test-data/unit/check-async-await.test]: Yep, I suspect these will be more about specific message cleanup than anything else; I'll file an issue with the list of things you've spotted, so I don't lose track of them.

[the remark this answers is not in the document]
Comment (OddBloke) [reply test-data/unit/check-async-await.test]: I've opened #3887 for this.

File: mypy/messages.py [hunk 8]
```diff
@@ -184,69 +184,27 @@ def warn(self, msg: str, context: Context, file: Optional[str] = None,
...
 
-    def format_simple(self, typ: Type, verbosity: int = 0) -> str:
-        """Convert simple types to string that is suitable for error messages.
+        This method returns a string appropriate for unmodified use in error
+        messages; this means that it will be quoted in most cases.  If
+        modification of the formatted string is required, callers should use
+        .format_bare.
+        """
+        ret = self.format_bare(typ, verbosity)
+        if ret in ['Module', 'overloaded function']:
```
[the remark this answers is not in the document]
Comment (OddBloke) [reply mypy/messages.py]: I've added the first two, but don't have time to add the union and tuple ones right now.

**Comment by OddBloke** [comment]
@ilevkivskyi So I'm looking at stripping quotes for union types, and the test I'm writing has this output now:

```
Argument 1 to "takes_int" has incompatible type "union type (6 items)"; expected "int"
```

but would have:

```
Argument 1 to "takes_int" has incompatible type union type (6 items); expected "int"
```

which I find quite hard to parse. (The tuple formatting doesn't have the same problem, because it doesn't repeat the word "type" (it's just `tuple(length N)`).

I wonder if we should leave the union representation in quotes and then perhaps improve it in a different PR? (I was thinking `<union type: N items>` and `<tuple: length N>` (or even `<tuple: N items>` if we want it to be consistent), which I think are easier to parse).

**Comment by OddBloke** [comment]
(I've pushed up a commit which does it just for tuples for now.)

**Comment by ilevkivskyi** [comment]
Actually I like the idea of ``<union: N items>`` and ``<tuple: N items>``, but again it is better for a separate PR.

**Comment by OddBloke** [comment]
\o/ Thanks!
